File: backend/app/schemas/project.py
from pydantic import BaseModel, ConfigDict, Field

from app.domain.source_file import SourceLanguage


class ProjectCreateRequest(BaseModel):
    root_dir: str = Field(
        ...,
        description="Relative sub-folder path inside the branch repository, e.g. '.' or 'backend/'",
    )
    language: SourceLanguage = Field(
        ...,
        description="Primary language of the sub-project, e.g. 'python', 'typescript'",
    )


class ProjectUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    root_dir: str | None = None
    language: SourceLanguage | None = None


class ProjectResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    branch_id: int
    workspace_id: int | None = None
    root_dir: str
    language: SourceLanguage

File: frontend/src/api/types/project.ts
/** Canonical languages supported by the indexing pipeline (backend SourceLanguage enum). */
export type SourceLanguage = 'python' | 'javascript' | 'typescript';

export const SOURCE_LANGUAGES: SourceLanguage[] = ['python', 'javascript', 'typescript'];

/**
 * A "Project" in backend vocabulary is a sub-project / SCIP indexing target
 * nested under a Branch (a root_dir + language pair) — NOT the same concept
 * as this app's Workspace. Labeled "Sub-project" in the UI to avoid confusion.
 */
export interface ProjectResponse {
  id: number;
  branch_id: number;
  workspace_id?: number | null;
  root_dir: string;
  language: SourceLanguage;
}

export interface ProjectCreateRequest {
  root_dir: string;
  language: SourceLanguage;
}
